Add polling module discovery for Fuchsia

jonahwilliams opened the pull request:
Fixes https://github.com/flutter/flutter/issues/24938

Commit:
add polling module discovery

Commit:
Update fuchsia_device.dart

Commit:
Merge branch 'master' of github.com:flutter/flutter into wait_for_attach

Commit:
Merge branch 'wait_for_attach' of github.com:jonahwilliams/flutter into wait_for_attach

Commit:
Update fuchsia_device.dart

Commit:
Merge branch 'master' of github.com:flutter/flutter into wait_for_attach

Commit:
remove extra log statement

tvolkert commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -288,6 +300,79 @@ class FuchsiaDevice extends Device {
+
+  void dispose() {
+    if (!_foundUri.isCompleted) {
+      _status.cancel();
```
`_status?.cancel();`

tvolkert replied:
... and then set it to `null`

Commit:
Update fuchsia_device.dart

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -288,6 +300,79 @@ class FuchsiaDevice extends Device {
+  Status _status;
+  bool _locked = false;
+
+  Future<Uri> get uri {
```
Maybe to avoid unnecessary awaits:
```dart
Uri _uri;
FutureOr<Uri> get uri {
  if (_uri != null) return _uri;
  ...
  return _foundUri.future.then((uri) {
    _uri = uri;
    return uri;
  });
}
```

jonahwilliams replied:
Done

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -288,6 +300,80 @@ class FuchsiaDevice extends Device {
+  }
+
+  void dispose() {
+    if (!_foundUri.isCompleted) {
```
Do you want to cancel the polling timer here, too?

jonahwilliams replied:
Done

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -288,6 +300,80 @@ class FuchsiaDevice extends Device {
+          service = await VMService.connect(uri);
+          _ports[port] = service;
+        } on SocketException catch (err) {
+          printTrace('Failed to connect to $localPort: $err');
```
Do you need to return here? It looks like you'll hit a null pointer exception otherwise.

jonahwilliams replied:
Added a continue, since it might not be fatal

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -288,6 +300,80 @@ class FuchsiaDevice extends Device {
+  Future<void> _findIsolate(Timer timer) async {
+    // Prevent a timer callback from firing while there is already an operation
+    // pending
+    if (_locked) {
```
Maybe use a regular timer with an explicit reschedule instead. Then you don't have to worry about this being re-entrant.

jonahwilliams replied:
Done

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -288,6 +300,80 @@ class FuchsiaDevice extends Device {
+  FuchsiaIsolateDiscoveryProtocol  getIsolateDiscoveryProtocol(String isolateName) => FuchsiaIsolateDiscoveryProtocol(this, isolateName);
+}
+
+class FuchsiaIsolateDiscoveryProtocol {
```
There's some non-trivial logic in here. Are the tests of 'attach' enough to cover the new polling code?

jonahwilliams replied:
No, added some test cases

Commit:
Merge branch 'master' of github.com:flutter/flutter into wait_for_attach

Commit:
address comments and add test

Commit:
address comments

jonahwilliams commented:
Updated polling to use a non-periodic timer which is rescheduled.  Added configurable max iterations and vmService connector for testing.

Added test cases to verify finding a matching isolate and two non matching cases

zanderso commented on packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:
```diff
@@ -278,6 +295,95 @@ class FuchsiaDevice extends Device {
+class FuchsiaIsolateDiscoveryProtocol {
+  FuchsiaIsolateDiscoveryProtocol(this._device, this._isolateName, [
+    this._vmServiceConnector = _kDefaultFuchsiaIsolateDiscoveryConnector,
+    this._maxIterations = -1,
```
Maybe leave a TODO to wire this up to something? Or remove it...

jonahwilliams replied:
This is a necessary evil for testing, but I don't think we want wire it up to the tool otherwise it becomes another timeout essentially.

jonahwilliams replied:
I could instead change this to a run once flag instead of max iter.

Commit:
replace max iter with pollOnce flag
